pkg/tecdsa/dkg/preparams_storage.go

```diff
@@ -0,0 +1,158 @@
+package dkg
+
+import (
+	"encoding/hex"
+	"fmt"
+	"sort"
+	"sync"
+
+	"github.com/ethereum/go-ethereum/crypto"
+	"github.com/ipfs/go-log/v2"
+
+	"github.com/keep-network/keep-common/pkg/persistence"
+	"github.com/keep-network/keep-core/pkg/generator"
+)
+
+const (
+	dirName = "preparams"
+)
+
+// PersistedPreParams is an alias for Persisted PreParams used in generator.Persistence
+// interface implementation.
+type PersistedPreParams = generator.Persisted[PreParams]
+
+type preParamsStorage struct {
+	// mutex is a single struct-wide lock that ensures all functions
+	// of the storage are thread-safe.
+	mutex sync.Mutex
+
+	persistence persistence.BasicHandle
+	logger      log.StandardLogger
+}
+
+func newPreParamsStorage(
+	persistence persistence.BasicHandle,
+	logger log.StandardLogger,
+) preParamsStorage {
+	return preParamsStorage{
+		persistence: persistence,
+		logger:      logger,
+	}
+}
+
+// Saves provided PreParams to the storage.
+func (p *preParamsStorage) Save(pp *PreParams) (*PersistedPreParams, error) {
+	p.mutex.Lock()
+	defer p.mutex.Unlock()
+
+	ppBytes, err := pp.Marshal()
+	if err != nil {
+		return nil, fmt.Errorf("marshalling of the preparams failed: [%v]", err)
+	}
+
+	fileName := fmt.Sprintf(
+		"pp_%d_%s",
+		// Use timestamp in the filename so that when the data are read from the
+		// disk the First In, First Out algorithm applies.
+		pp.creationTimestamp.UnixMilli(),
+		// Add part of the hash for an ultra unlikely scenario that two saves
+		// happen in the exactly the same millisecond.
+		hex.EncodeToString(crypto.Keccak256(ppBytes))[:7],
```

Comment:
Just to confirm, we want to take the first 7 bytes of the hash right?

Reply:
Yes, just a part of the hash.